core/chains/cosmos/chain.go

```diff
@@ -0,0 +1,168 @@
+var _ adapters.Chain = (*chain)(nil)
+
+type chain struct {
+	utils.StartStopOnce
+	id           string
+	cfg          coscfg.Config
+	cfgImmutable bool // toml config is immutable
+	txm          *cosmostxm.Txm
+	orm          types.ORM
+	lggr         logger.Logger
+}
+
+func newChain(id string, cfg coscfg.Config, db *sqlx.DB, ks keystore.Cosmos, logCfg pg.QConfig, eb pg.EventBroadcaster, orm types.ORM, lggr logger.Logger) (*chain, error) {
+	lggr = lggr.With("cosmosChainID", id)
+	var ch = chain{
+		id:   id,
+		cfg:  cfg,
+		orm:  orm,
+		lggr: lggr.Named("Chain"),
+	}
+	tc := func() (cosmosclient.ReaderWriter, error) {
+		return ch.getClient("")
+	}
+	gpe := cosmosclient.NewMustGasPriceEstimator([]cosmosclient.GasPricesEstimator{
+		cosmosclient.NewClosureGasPriceEstimator(func() (map[string]sdk.DecCoin, error) {
+			return map[string]sdk.DecCoin{
+				"uatom": sdk.NewDecCoinFromDec("uatom", cfg.FallbackGasPriceUAtom()),
+			}, nil
+		}),
+	}, lggr)
+	ch.txm = cosmostxm.NewTxm(db, tc, *gpe, ch.id, cfg, ks, lggr, logCfg, eb)
+
+	return &ch, nil
+}
+
+func (c *chain) Name() string {
+	return c.lggr.Name()
+}
+
+func (c *chain) ID() string {
+	return c.id
+}
+
+func (c *chain) Config() coscfg.Config {
+	return c.cfg
+}
+
+func (c *chain) UpdateConfig(cfg *db.ChainCfg) {
+	if c.cfgImmutable {
+		c.lggr.Criticalw("TOML configuration cannot be updated", "err", v2.ErrUnsupported)
+		return
+	}
+	c.cfg.Update(*cfg)
+}
+
+func (c *chain) TxManager() adapters.TxManager {
+	return c.txm
+}
+
+func (c *chain) Reader(name string) (cosmosclient.Reader, error) {
+	return c.getClient(name)
+}
+
+// getClient returns a client, optionally requiring a specific node by name.
+func (c *chain) getClient(name string) (cosmosclient.ReaderWriter, error) {
+	//TODO cache clients?
+	var node db.Node
+	if name == "" { // Any node
+		nodes, cnt, err := c.orm.NodesForChain(c.id, 0, math.MaxInt)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get nodes")
+		}
+		if cnt == 0 {
+			return nil, errors.New("no nodes available")
+		}
+		// #nosec
+		node = nodes[rand.Intn(len(nodes))]
+	} else { // Named node
+		var err error
+		node, err = c.orm.NodeNamed(name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get node named %s", name)
+		}
+		if node.CosmosChainID != c.id {
+			return nil, fmt.Errorf("failed to create client for chain %s with node %s: wrong chain id %s", c.id, name, node.CosmosChainID)
+		}
+	}
+	client, err := cosmosclient.NewClient(c.id, node.TendermintURL, DefaultRequestTimeout, c.lggr.Named("Client-"+name))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create client")
+	}
+	c.lggr.Debugw("Created client", "name", node.Name, "tendermint-url", node.TendermintURL)
+	return client, nil
+}
+
+// Start starts cosmos chain.
+func (c *chain) Start(ctx context.Context) error {
+	return c.StartOnce("Chain", func() error {
+		c.lggr.Debug("Starting")
+		//TODO dial client?
```

Comment:
Is this a new comment? or copied from another impl?

Reply:
I think we can remove it. It's an optional thing that doesn't provide that much benefit.